pkgtools/R2pkg/files/R2pkg.8
.\"	$NetBSD: R2pkg.8,v 1.4 2017/07/18 18:44:06 wiz Exp $
.\"
.\" Copyright (c) 2014,2015,2016,2017
.\"	Brook Milligan.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the author nor the names of any contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"
.Dd February 23, 2017
.Dt R2PKG 8
.Os
.Sh NAME
.Nm R2pkg
.Nd create a pkgsrc package for an R package
.Sh SYNOPSIS
.Nm
.Op Fl DehqruVv
.Op Fl E Ar editor
.Op Fl R Ar dependency_file
.Op Ar package
.Sh DESCRIPTION
.Nm
takes the name of an R
.Ar package
that should be turned into a
.Xr pkgsrc 7
package in the current directory.
The generated package includes a
.Pa Makefile
and a mostly complete set of files to get the package going.
However,
the files should be reviewed and perhaps adjusted manually after
.Nm
has finished its job.
To help with this, the original versions of files created by
.Nm
are preserved for reference.
.Pp
.Nm
is intended to help the process of creating a package but is not
intended to fly on autopilot.
.Pp
The
.Pa Makefile
contains all of the variables required for an R package, with
information being taken from the package's
.Pa DESCRIPTION
file on CRAN, which is automatically fetched by
.Nm .
Likewise, the
.Pa DESCR
file is populated with information from the same file.
After
.Pa Makefile
and
.Pa DESCR
are completed,
.Nm
fetches the package and computes its checksum via the
.Ic makesum
target.
.Pp
If the user does not supply the
.Ar package
on the command line, then
.Nm
will prompt for it.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl D
Write the package's description into
.Pa DESCRIPTION .
This may be useful for verifying the information obtained from CRAN.
However, this file must be removed as it is not part of a
.Xr pkgsrc 7
package.
.It Fl E Ar editor
Use
.Ar editor
instead of the user's default editor.
.It Fl e
Do not edit Makefile and DESCR.
.It Fl h
Produce a short help message.
.It Fl q
Do not produce status messages along the way.
.It Fl R Ar dependency_file
Process dependency packages recursively using
.Pa dependency_file
to record dependency information.
This option is intended for
internal use only to implement recursion over dependencies.
Instead, the
.Fl r
option should be used to select recursion.
.It Fl r
Process dependency packages recursively.
A
.Pa depends
file is created that lists the dependencies in
.Xr tsort 1
order.
This is likely the best order for testing the resulting
dependency packages.
In particular, a useful strategy for creating
packages recursively is to (i) run
.Nm
with the
.Fl r
option, (ii) fix all the leaf packages, (iii) rerun
.Nm
with the
.Fl r
option, and (iv) continue until all packages are created.
Repeating runs of
.Nm
once the leaf dependencies are moved into the correct categories will
allow the program to capture the correct dependencies in each
.Pa Makefile .
.It Fl u
Update the
.Xr pkgsrc 7
package.
.Nm
must be run in an R package directory, which is assumed to be named as
.Pa R-package .
The name of the package to be updated will be taken from the directory
name and need not be given on the command line.
.It Fl V
Print the version.
.It Fl v
Generate verbose output.
This may be repeated to increase the verbosity.
.El
.Sh ENVIRONMENT
The following environment variables affect the execution of
.Nm :
.Bl -tag -width BLOCKSIZE
.It Ev PKGEDITOR , EDITOR
Editor used for editing
.Pa Makefile
and
.Pa DESCR .
.It Ev CRAN_URL
The URL to the CRAN mirror for downloading the package's DESCRIPTION
file.
By default,
.Pa ftp://cran.r-project.org .
.El
.Sh EXAMPLES
Use the following commands to create a
.Xr pkgsrc 7
package for the R package
.Ar foo :
.Dl cd pkgsrc/math
.Dl mkdir R-foo
.Dl cd R-foo
.Dl R2pkg foo
Remember to test the resulting package and remove any extra files
created by
.Nm .
.Sh SEE ALSO
.Xr pkgsrc 7
.Sh AUTHORS
.Nm
was written by
.An Brook Milligan Aq Mt brook@NetBSD.org .
.Sh BUGS
.Nm
does not create a finished package; instead, it facilitates the task.
However, the resulting files must be reviewed and possibly edited by
hand.
Although dependencies are included in the generated
.Pa Makefile ,
they are not fully analyzed and likely have incorrect category
information.
Packages created recursively for dependencies are placed
in a newly created, nonstandard
.Pa pkgsrc/R
category directory and must be moved to appropriate category
directories within
.Xr pkgsrc 7
and checked for correctness.


doc/NEWS
# $NetBSD: NEWS,v 1.11 2014/06/21 16:01:17 alnsn Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

Lua multiversion support [alnsn 2014-05-03]
	Lua packages lang/lua51 and lang/lua52 have been modified
	to coexist without conflicts, infrastructure for supporting
	multiple versions of Lua has been put in place and all
	dependent packages have been adapted to the new infrastructure.

SMF support is introduced [jperkin 2014-03-11]

Unprivileged staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"
	to invoke this functionality. Set PKG_DESTDIR_SUPPORT to "none"
	to revert to direct installation.

buildlink3.mk layout changed [joerg 2009-03-20]
	The buildlink3.mk files have been reorganized to reduce
	redundant parsing for deeply nested buildlink3 trees.
	The createbuildlink package has been updated to create
	the new format.

PLIST_VARS introduced [jlam 20080413]
	PLIST_VARS is a simpler way to do conditional PLIST entries.
	The old method of doing e.g. PLIST_SUBST+= FOO="@comment " can
	now be done by unconditionally setting PLIST_VARS+=foo and
	conditionally setting PLIST.foo=yes. This will result in substitution
	of ${PLIST.foo} in the PLIST with either "" or "@comment ".

SKIP_PORTABILITY_CHECK is obsolete [rillig 20080220]
	After being deprecated for about 16 months, this early misdesign
	has finally been obsoleted and removed from pkgsrc.

Check for stripped/unstripped binaries [rillig 20080213]
	A new check has been added in mk/check/check-stripped.mk to
	assure that the installed binaries conform to the setting of the
	INSTALL_UNSTRIPPED variable. It is not enabled by default due to
	lack of testing.

OpenSSL update [tnn 20080118]
	OpenSSL in pkgsrc switched from the 0.9.7 to the 0.9.8 branch. To
	accomodate for this, 596 packages dependent on openssl had their
	revisions bumped.

pkgsrc-2007Q4 released [agc 20080105]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2008/01/05/0000.html

Simplifying meta-packages [rillig 20071219]
	The variable META_PACKAGE was introduced. This simplifies creation
	and maintenance of packages that do not install files.

Replace a package with different name [gdt 20071107]
	The variable OLDNAME was introduced for "make replace" to
	replace a package with a name other than the new one, via
	"make replace OLDNAME=oldpkgname".  Read the announcement at:
	http://mail-index.netbsd.org/tech-pkg/2007/11/07/0002.html
	
pkgsrc-2007Q3 released [agc 20071015]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/10/15/0000.html

Features framework [jlam 20070907]
	A "system features" framework has been added. It will be used to
	automatically supply missing "basic" headers and libraries for old
	systems. This is supported through the USE_FEATURES variable.

X.org 6.9 removed [joerg 20070817]
	monolithic X.org 6.9 was removed from pkgsrc. It was obsoleted by the
	modular X.org (7.x) packages. Transitioning users should remove all old
	X.org packages and set X11_TYPE=modular in mk.conf.

Emulator framework [jlam 20070729]
	A new emulator framework has been added. This will handle all
	binary-only packages that require emulation on the native OS.
	Please see pkgsrc/mk/emulator/README.

audit-packages merged with pkg_install [adrianp 20070714]
	The utility for auditing installed packages for security problems,
	audit-packages, has been rewritten in C. It is now part of the
	pkgtools/pkg_install package.
	
pkgsrc-2007Q2 released [agc 20070701]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/07/01/0000.html

New framework for bulk builds [joerg 20070619]
	pbulk, the new automated pkgsrc bulk build framework was imported.

pkgsrc-2007Q1 released [agc 20070419]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/04/19/0000.html

pkgsrc-2006Q4 released [agc 20070104]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/01/04/0000.html

modular X.org imported [joerg 20061114]
	The modular X.org packages were imported.

pkgsrc-2006Q3 released [agc 20061001]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2006/10/01/0000.html


pkgtools/R2pkg/files/R2pkg.sh
#!/bin/sh
# R2pkg
#
# Copyright (c) 2014,2015,2016,2017
#	Brook Milligan.  All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. Neither the name of the author nor the names of any contributors
#    may be used to endorse or promote products derived from this software
#    without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
#
# Create an R package in the current directory
#
NAME="R2pkg"
VERS="@VERS@"

USAGE="${NAME} [-DVehqrv] [-E editor] [-R dependency_file] [package] -- create an R package for pkgsrc"

: ${CRAN_URL:=ftp://cran.r-project.org}
: ${PKGEDITOR:=${EDITOR:=vi}}

# Substituted by pkgsrc at pre-configure time.
MAKE=@MAKE@
EDIT=1
QUIET=0
RECURSIVE=0
TOP_LEVEL=1
UPDATE=0
VERBOSE=0

DESCRIPTION_CONNECTION=connection

while getopts DE:R:Vehqruv f
do
    case $f in
	D) DESCRIPTION=yes; DESCRIPTION_CONNECTION="'DESCRIPTION'"; ARGS="${ARGS} -D";;
	E) PKGEDITOR=${OPTARG}; ARGS="${ARGS} -E ${PKGEDITOR}";;
	R) DEPENDENCY_LIST=${OPTARG}; RECURSIVE=1; TOP_LEVEL=0; ARGS="${ARGS} -R ${DEPENDENCY_LIST}";;
	V) echo "${NAME} v${VERS}"; exit 0;;
	e) EDIT=0; ARGS="${ARGS} -e";;
	h) echo ${USAGE}; exit 0;;
	q) QUIET=1; ARGS="${ARGS} -q";;
	r) RECURSIVE=1; RECURSIVE_MESSAGE=1; ARGS="${ARGS} -r";;
	u) UPDATE=1;;
	v) VERBOSE=$((${VERBOSE}+1)); ARGS="${ARGS} -v";;
        \?) echo ${USAGE}; exit 1;;
    esac
done
shift `expr ${OPTIND} - 1`

if [ ${UPDATE} -eq 1 ]; then
    RPKG=$(echo $(basename $(pwd)) | sed -e 's/^R-//');
elif [ ${#} -eq 0 ]; then
    read -p "package: " RPKG TAIL
    if [ "X${TAIL}" != "X" ]; then
	echo "Error: multiple package names given."
	echo ${USAGE}
	exit 1
    fi
elif [ ${#} -eq 1 ]; then
    RPKG=${1}
else
    echo ${USAGE}
    exit 1
fi

R_FILE=${TMPDIR}/R2pkg.$$.R

if [ ${TOP_LEVEL} -eq 1 ]; then
    BANNER_MSG="===> Creating R package: R-${RPKG}"
else
    BANNER_MSG="===> Creating dependency package: R-${RPKG}"
fi
if [ "X${DEPENDENCY_LIST}" = "X" ]; then
    DEPENDENCY_LIST=${TMPDIR}/R2pkg.depends.$$
    ARGS="${ARGS} -R ${DEPENDENCY_LIST}"
fi
if [ ${QUIET} -eq 1 ]; then
    STDOUT_MAKESUM=">/dev/null"
    STDOUT_EXTRACT=">/dev/null"
    QUIET_CURL="TRUE"
    ECHO_BANNER=":"
    ECHO_DONE=":"
    ECHO_FETCH=":"
    ECHO_EXTRACT=":"
    if [ ${TOP_LEVEL} -eq 0 ]; then
	ECHO=":"
    fi
elif [ ${VERBOSE} -eq 0 ]; then
    STDOUT_MAKESUM=">/dev/null"
    STDOUT_EXTRACT=">/dev/null"
    QUIET_CURL="TRUE"
    ECHO_BANNER="echo"
    ECHO_DONE=":"
    ECHO_FETCH=":"
    ECHO_EXTRACT="echo"
elif [ ${VERBOSE} -eq 1 ]; then
    STDOUT_MAKESUM=">/dev/null"
    STDOUT_EXTRACT=">/dev/null"
    QUIET_CURL="TRUE"
    ECHO_BANNER="echo"
    ECHO_DONE="echo"
    ECHO_FETCH="echo"
    ECHO_EXTRACT="echo"
else
    STDOUT_MAKESUM=""
    STDOUT_EXTRACT=""
    QUIET_CURL="FALSE"
    ECHO_BANNER="echo"
    ECHO_DONE="echo"
    ECHO_FETCH="echo"
    ECHO_EXTRACT="echo"
fi

CRAN_PACKAGES=pub/R/web/packages
RPKG_DESCRIPTION_URL=${CRAN_URL}/${CRAN_PACKAGES}/${RPKG}/DESCRIPTION

check_for_R ()
{
    R_CMD="Rscript --no-save /dev/null"
    eval ${R_CMD}
    if [ ${?} -ne 0 ]; then
	echo "ERROR: math/R package is not installed."
	exit 1
    fi
}

preserve_original_files ()
{
    [ -f DESCR ]    && mv DESCR DESCR.orig
    [ -f Makefile ] && mv Makefile Makefile.orig
    [ -f distinfo ] && mv distinfo distinfo.orig
}

make_package ()
{
    R_CMD="Rscript --no-save ${R_FILE}"
    cat << EOF > ${R_FILE}
trim.space <- function(s) gsub('[[:space:]]','',s)
trim.blank <- function(s) gsub('[[:blank:]]','',s)
one.space <- function(s) gsub('[[:blank:]]+',' ',s)
one.line <- function(s) gsub('\n',' ',s)
pkg.vers <- function(s) gsub('_','.',s)
field <- function(key,value) paste(key,'=\t',value,sep='')

licenses <- list()
licenses[['ACM']]           <- 'acm-license'
licenses[['APACHE']]        <- 'apache-2.0'
licenses[['ARTISTIC']]      <- 'artistic-2.0'
licenses[['BSD-2']]         <- '2-clause-bsd'
licenses[['GPL-2']]         <- 'gnu-gpl-v2'
licenses[['GPL-3']]         <- 'gnu-gpl-v3'
licenses[['GPL (>= 2)']]    <- 'gnu-gpl-v2'
licenses[['GPL-2 | GPL-3']] <- 'gnu-gpl-v2 OR gnu-gpl-v3'
licenses[['LGPL-2']]        <- 'gnu-lgpl-v2'
licenses[['LGPL-2.1']]      <- 'gnu-lgpl-v2.1'
licenses[['LGPL (>= 2)']]   <- 'gnu-lgpl-v2'
licenses[['LUCENT']]        <- 'lucent'
licenses[['MIT']]           <- 'mit'
licenses[['POSTGRESQL']]    <- 'postgresql-license'

paste2 <- function(s1,s2)
{
  if (is.na(s1) && is.na(s2)) return ('')
  if (is.na(s1) && !is.na(s2)) return (s2)
  if (!is.na(s1) && is.na(s2)) return (s1)
  if (!is.na(s1) && !is.na(s2)) return (paste(s1,s2))
}

todo.license <- function(s)
{
  if (is.null(licenses[[s]]))
    todo <- '# TODO: LICENSE'
  else
    todo <- 'LICENSE'
  todo
}

pkgsrc.license <- function(s)
{
  license <- licenses[[s]]
  if (is.null(license))
    license <- s
  license
}

package <- function(s) field('R_PKGNAME',one.line(s))
version <- function(s) field('R_PKGVER',one.line(s))
comment <- function(s) field('COMMENT',one.line(s))
license <- function(s) field(todo.license(s),pkgsrc.license(s))

categories <- function() paste('CATEGORIES=',paste(basename(dirname(getwd())),'R'),sep='	')
description <- function(s) strwrap(s,width=71)

filter.imports <- function(s)
{
  base.packages <- c('R','MASS','Matrix','Rcpp','cluster','grDevices','graphics','grid','foreign','lattice','nlme','methods','nnet','parallel','rpart','stats','survival','tools','utils')
  for (pkg in base.packages)
    {
      re.pkg <- paste('^',pkg,sep='')
      s <- s[!grepl(re.pkg,s)]
    }
  s
}

make.imports <- function(s1,s2)
{
  s <- paste2(s1,s2)
  s <- gsub('([[:alnum:]]+)[[:blank:]]+(\\\([^\\\)]*\\\))?[[:blank:]]*,?','\\\\1 \\\\2,',s)
  imports <- na.omit(unlist(strsplit(s,',[[:blank:]]*')))
  imports <- trim.space(imports)
  imports <- filter.imports(imports)
  imports
}

make.dependency <- function(s)
{
  s <- gsub('\\\\)','',s)
  s <- unlist(strsplit(s,'\\\\('))
  s
}

depends <- function(s1,s2)
{
  imports <- make.imports(s1,s2)
  DEPENDS <- list()
  if (length(imports) > 0)
    {
      for (i in 1:length(imports))
        {
          dependency <- make.dependency(imports[i])
          depends <- dependency[1]
          depends.pkg <- Sys.glob(paste('../../*/R-',depends,sep=''))
          if (length(depends.pkg) != 1) # a unique dependency cannot be found
            {
              if (${RECURSIVE})
                {
                  dependency.dir <- paste('../../R/R-',depends,sep='')
                  dir.create(path=dependency.dir,recursive=TRUE)
                  error <- system(paste('(cd',dependency.dir,'&& R2pkg.sh ${ARGS}',depends,')'))
                  if (error != 0)
                    file.remove(dependency.dir)
                }
            }
          depends.pkg <- Sys.glob(paste('../../*/R-',depends,sep=''))
          if (length(depends.pkg) == 1) # a unique dependency found
            {
              depends.pkg.fullname <- system(paste('cd',depends.pkg,'&& bmake show-var VARNAME=PKGNAME'),intern=TRUE)
              depends.pkg.name <- sub('^(.*)-([^-]*)$','\\\\1',depends.pkg.fullname)
              depends.pkg.vers <- sub('^(.*)-([^-]*)$','\\\\2',depends.pkg.fullname)
              if (length(dependency) == 2)
                depends.vers <- dependency[2]
              else
                depends.vers <- paste('>=',depends.pkg.vers,sep='')
              depends.line <- paste('DEPENDS+=\tR-',depends,depends.vers,':',depends.pkg,sep='')
              if (length(dependency) == 2)
                depends.line <- paste(depends.line,'	# XXX - found ',depends.pkg.fullname,' (',depends.pkg,')',sep='')
            }
          else
            {
              depends.vers <- ifelse(length(dependency) == 2, dependency[2], '>=???')
              depends.vers <- trim.space(depends.vers)
              depends.line <- paste('DEPENDS+=\tR-',depends,depends.vers,':../../R/R-',depends,sep='')
            }
          DEPENDS <- list(DEPENDS,depends.line)
          new.depends.pkg <- Sys.glob(paste('../../R/R-',depends,sep=''))
          if (length(new.depends.pkg) > 0)
            system(paste('echo',depends,'${RPKG} >> ${DEPENDENCY_LIST}'))
        }
    }
  if (length(DEPENDS) > 0)
    DEPENDS <- append(DEPENDS,'')
  DEPENDS
}

use.languages <- function(s1,s2)
{
  USE_LANGUAGES <- list()
  s <- paste(s1,s2)
  Rcpp <- grepl('Rcpp',s)
  if (Rcpp)
    USE_LANGUAGES <- append(USE_LANGUAGES,list('USE_LANGUAGES+=	c++'))
  if (length(USE_LANGUAGES) > 0)
    USE_LANGUAGES <- append(USE_LANGUAGES,'')
  USE_LANGUAGES
}

buildlink <- function(s1,s2)
{
  BUILDLINK <- list()
  s <- paste(s1,s2)
  Rcpp <- grepl('Rcpp',s)
  if (Rcpp)
    BUILDLINK <- append(BUILDLINK,'.include "../../devel/R-Rcpp/buildlink3.mk"')
  BUILDLINK
}

copy.description <- function(connection)
{
  description <- readLines(connection)
  writeLines(description,con='DESCRIPTION')
}

error <- download.file(url='${RPKG_DESCRIPTION_URL}',destfile='DESCRIPTION',quiet=${QUIET_CURL},method='curl')
if (error)
  quit(status=error)

metadata <- read.dcf(file='DESCRIPTION', fields=c('Package','Version','Title','Description','License','Imports','Depends'))
 
CVS               <- '# \$NetBSD\$'
CATEGORIES        <- categories()
MASTER.SITES      <- 'MASTER_SITES=	\${MASTER_SITE_R_CRAN:=contrib/}'
MAINTAINER        <- 'MAINTAINER=	pkgsrc-users@NetBSD.org'
HOMEPAGE          <- 'HOMEPAGE=	\${R_HOMEPAGE_BASE}/${RPKG}/'
COMMENT           <- comment(metadata[3])
LICENSE           <- license(metadata[5])
R_PKGNAME         <- package(metadata[1])
R_PKGVER          <- version(metadata[2])
USE_LANGUAGES     <- use.languages(metadata[6],metadata[7])
DEPENDS           <- depends(metadata[6],metadata[7])
INCLUDE.R         <- '.include "../../math/R/Makefile.extension"'
INCLUDE.BUILDLINK <- buildlink(metadata[6],metadata[7])
INCLUDE.PKG       <- '.include "../../mk/bsd.pkg.mk"'

DESCR        <- description(metadata[4])

Makefile <- list()
Makefile <- append(Makefile,CVS)
Makefile <- append(Makefile,'')
Makefile <- append(Makefile,CATEGORIES)
Makefile <- append(Makefile,MASTER.SITES)
Makefile <- append(Makefile,'')
Makefile <- append(Makefile,MAINTAINER)
Makefile <- append(Makefile,HOMEPAGE)
Makefile <- append(Makefile,COMMENT)
Makefile <- append(Makefile,LICENSE)
Makefile <- append(Makefile,'')
Makefile <- append(Makefile,R_PKGNAME)
Makefile <- append(Makefile,R_PKGVER)
Makefile <- append(Makefile,'')
Makefile <- append(Makefile,USE_LANGUAGES)
Makefile <- append(Makefile,DEPENDS)
Makefile <- append(Makefile,INCLUDE.R)
Makefile <- append(Makefile,INCLUDE.BUILDLINK)
Makefile <- append(Makefile,INCLUDE.PKG)
Makefile <- paste(unlist(Makefile),collapse='\n')

write(Makefile,'Makefile')
write(DESCR,'DESCR')
EOF
    eval ${R_CMD}
    retval=${?}
    if [ ${retval} -ne 0 ]; then
	echo "ERROR: making ${RPKG} package failed."
    fi
    return ${retval}
}

edit_Makefile ()
{
    if [ ${EDIT} -ne 0 -a -s Makefile ]; then
	${PKGEDITOR} Makefile
    fi
}

edit_DESCR ()
{
    if [ ${EDIT} -ne 0 -a -s DESCR ]; then
	${PKGEDITOR} DESCR
    fi
}

create_distinfo ()
{
    ${ECHO_FETCH} "==> Fetching R-${RPKG} ..."
    MAKE_CMD="${MAKE} makesum ${STDOUT_MAKESUM}"
    eval ${MAKE_CMD}
    return ${?}
}

extract ()
{
    ${ECHO_EXTRACT} "==> Extracting R-${RPKG} ..."
    MAKE_CMD="${MAKE} extract ${STDOUT_EXTRACT}"
    eval ${MAKE_CMD}
}

cleanup ()
{
    [ "X${DESCRIPTION}" != "X" ] || rm -f DESCRIPTION
    if [ -f DESCR.orig ] && cmp -s DESCR.orig DESCR; then
	mv DESCR.orig DESCR
    fi
    if [ -f Makefile.orig ] && cmp -s Makefile.orig Makefile; then
	mv Makefile.orig Makefile
    fi
    if [ -f distinfo.orig ] && cmp -s distinfo.orig distinfo; then
	mv distinfo.orig distinfo
    fi
    rm -f ${R_FILE}
}

messages ()
{
    if [ ${QUIET} -eq 0 -a ${TOP_LEVEL} -ne 0 ]; then
	cat << EOF

Please do not forget the following:
- check DESCR and edit as necessary.
- check Makefile:
  o verify the COMMENT.
  o verify the LICENSE.
  o verify the DEPENDS, especially the categories.
EOF
	[ "X${DESCRIPTION}" != "X" ] && echo "- remove DESCRIPTION."
	if [ ${RECURSIVE} -ne 0 ]; then
	    cat << EOF

Recursive packages may have been created in ../../R; please do the following:
- edit each Makefile as follows (in addition to following the notes above):
  o move recursively created packages to the appropriate category.
  o fix the category in Makefile.
  o fix the category of any dependencies.
  o remove any extraneous dependencies.
EOF
	    if [ -s ${DEPENDENCY_LIST} ]; then
		tsort ${DEPENDENCY_LIST} > depends
		echo "- It may be useful to test these packages in the following order:"
		awk 'BEGIN{printf(" ")} {printf(" R-%s",$0)}' depends && echo
		[ ${TOP_LEVEL} -eq 0 ] || rm -f ${DEPENDENCY_LIST}
	    fi
	fi
    fi
}

${ECHO_BANNER} "${BANNER_MSG} ..."
check_for_R
preserve_original_files
make_package
error=${?}
if [ ${error} -eq 0 ]; then
    edit_Makefile
    edit_DESCR
    create_distinfo
    error=${?}; [ ${error} -eq 0 ] || exit ${error}
    # extract
fi
cleanup
messages
${ECHO_DONE} "${BANNER_MSG}: done"
exit ${error}
